File: day19.py
import math
import re
from collections import deque
from functools import partial, reduce
from multiprocessing import Pool
from operator import mul
from pathlib import Path
from typing import Pattern

COST_RE: Pattern = re.compile(r"\d+\s[a-z]+")

test_data = Path("inputs/19_t.txt").read_text().splitlines()
real_data = Path("inputs/19.txt").read_text().splitlines()
real: bool = True

the_data = real_data if real else test_data


CostItem = tuple[int, int]
Cost = list[CostItem]
Costs = list[Cost]
GEODE = 3


def get_costs(bp_line: str) -> list:
    return [COST_RE.findall(_) for _ in bp_line.strip().split(".")[:-1]]


def parse_blueprint(blueprint: str) -> Costs:
    costs: Costs = []
    for i, recipes in enumerate(get_costs(blueprint)):
        costs.append([])
        for recipe in recipes:
            amt, rtype = recipe.split(" ")
            costs[i].append(
                (int(amt), ["ore", "clay", "obsidian"].index(rtype))
            )
    return costs


def wait_time(costs: list[CostItem], res: list[int], bots: list[int]) -> int:
    max_wait = 0
    for amt, resource_type in costs:
        if bots[resource_type] == 0:
            return 2**8
        max_wait = max(
            max_wait,
            math.ceil((amt - res[resource_type]) / bots[resource_type]),
        )
    return max_wait


def get_max_bots(costs: Costs) -> list[int]:
    max_bots = [0, 0, 0]
    for blueprint in costs:
        for (amt, rtype) in blueprint:
            max_bots[rtype] = max(max_bots[rtype], amt)
    return max_bots


def dfs(max_bots: list[int], costs: Costs, time: int) -> int:
    seen = set()
    queue = deque([([0, 0, 0, 0], [1, 0, 0, 0], time)])
    best = 0

    while queue:
        res, bots, time = queue.popleft()

        best = max(best, res[GEODE] + bots[GEODE] * time)

        for bot_type, blueprint in enumerate(costs):
            # Obs 1: do not build more than the max_bots number of bots
            # for each type (except geode)
            if bot_type != GEODE and bots[bot_type] >= max_bots[bot_type]:
                continue

            wait = wait_time(blueprint, res, bots)

            _t = time - wait - 1
            if _t <= 0:
                continue
            _bots = bots[:]
            _res = [old + rate * (wait + 1) for old, rate in zip(res, bots)]
            _bots[bot_type] += 1
            for amt, res_type in blueprint:
                _res[res_type] -= amt

            # Obs 2: throw away extra resources
            for res_type in range(3):
                _res[res_type] = min(_res[res_type], max_bots[res_type] * _t)

            _k = tuple([time, *_res, *_bots])
            if _k not in seen:
                queue.append((_res, _bots, _t))
                seen.add(_k)

    return best


def solve(line: str, start_time: int) -> int:
    m_costs = parse_blueprint(line)
    return dfs(get_max_bots(m_costs), m_costs, start_time)


with Pool() as pool:
    # Part 1
    ans1 = sum(
        (i + 1) * v
        for i, v in enumerate(pool.map(partial(solve, start_time=24), the_data))
    )
    print(ans1)

    # Part 2
    ans2 = reduce(mul, pool.map(partial(solve, start_time=32), the_data[:3]), 1)
    print(ans2)

File: 2022/src/bin/day19.rs
use std::{
    cmp::{max, min},
    collections::VecDeque,
    fs,
};

use ahash::HashSet;
use aoc_prelude::lazy_static;
use rayon::prelude::*;
use regex::Regex;

#[derive(Debug)]
struct CostItem {
    amount: usize,
    resource_type: usize,
}

type Bots = [usize; 4];
type Res = [usize; 4];

#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Debug)]
struct Wunit {
    res: Res,
    bots: Bots,
    time: usize,
}

const GEODE: usize = 3;

fn parse_blueprint(blueprint: &str) -> Vec<Vec<CostItem>> {
    lazy_static! {
        static ref COST_RE: Regex = Regex::new(r"\d+\s[a-z]+").unwrap();
        static ref RES: Vec<&'static str> = vec!["ore", "clay", "obsidian"];
    }
    let mut res = Vec::new();
    for recipe in blueprint.split(".") {
        let mut inner = Vec::new();
        for m in COST_RE.find_iter(recipe) {
            let mut parts = m.as_str().split(" ");
            let amount = parts
                .next()
                .map(|x| x.parse::<usize>().expect("invalid amount"))
                .unwrap();
            let resource_type = parts.next().unwrap();
            let resource_type = RES
                .iter()
                .position(|x| *x == resource_type)
                .expect("invalid resource");
            inner.push(CostItem {
                amount,
                resource_type,
            })
        }
        if !inner.is_empty() {
            res.push(inner);
        }
    }
    res
}

fn wait_time(costs: &Vec<CostItem>, res: Res, bots: Bots) -> usize {
    #[inline(always)]
    fn div_up(a: usize, b: usize) -> usize {
        (a + (b - 1)) / b
    }

    let mut max_wait = 0;
    for cost in costs {
        if bots[cost.resource_type] == 0 {
            return 2 << 8;
        }
        if res[cost.resource_type] <= cost.amount {
            max_wait = max(
                max_wait,
                div_up(
                    cost.amount - res[cost.resource_type],
                    bots[cost.resource_type],
                ),
            )
        }
    }
    max_wait
}

fn get_max_bots(costs: &[Vec<CostItem>]) -> [usize; 3] {
    let mut max_bots = [0, 0, 0];
    for blueprint in costs.iter() {
        for cost_item in blueprint.iter() {
            max_bots[cost_item.resource_type] =
                max(max_bots[cost_item.resource_type], cost_item.amount);
        }
    }
    max_bots
}

fn dfs(max_bots: [usize; 3], costs: &[Vec<CostItem>], time: usize) -> usize {
    let mut seen: HashSet<Wunit> = HashSet::default();
    let mut queue: VecDeque<Wunit> = VecDeque::default();
    queue.push_back(Wunit {
        res: [0, 0, 0, 0],
        bots: [1, 0, 0, 0],
        time,
    });

    let mut best = 0;

    while !queue.is_empty() {
        let wunit = queue.pop_front().unwrap();
        best = max(best, wunit.res[GEODE] + wunit.bots[GEODE] * wunit.time);

        for (bot_type, blueprint) in costs.iter().enumerate() {
            // Obs 1: do not build more than the max_bots number of bots
            // for each type (except geode)
            if bot_type != GEODE && wunit.bots[bot_type] >= max_bots[bot_type] {
                continue;
            }

            let wait = wait_time(blueprint, wunit.res, wunit.bots);

            if wait + 1 >= wunit.time {
                continue;
            }
            let _t = wunit.time - wait - 1;
            let mut _bots = wunit.bots;
            let mut _res_slice: [usize; 4] = wunit
                .res
                .iter()
                .zip(wunit.bots.iter())
                .map(|(old, rate)| old + rate * (wait + 1))
                .collect::<Vec<_>>()
                .try_into()
                .expect("nope");
            _bots[bot_type] += 1;

            for b in blueprint.iter() {
                _res_slice[b.resource_type] -= b.amount;
            }

            // Obs 2: throw away extra resources
            for res_type in 0..3 {
                _res_slice[res_type] = min(_res_slice[res_type], max_bots[res_type] * _t)
            }
            let _k = Wunit {
                res: _res_slice,
                bots: _bots,
                time: _t,
            };
            if !seen.contains(&_k) {
                seen.insert(_k.clone());
                queue.push_back(_k);
            }
        }
    }

    best
}

fn solve(line: &str, start_time: usize) -> usize {
    let costs = parse_blueprint(line);
    dfs(get_max_bots(&costs), &costs, start_time)
}

fn main() {
    let now = std::time::Instant::now();

    let file_path = "inputs/19.txt";
    let input: Vec<String> = fs::read_to_string(file_path)
        .expect("valid input file")
        .lines()
        .map(String::from)
        .collect();

    let p1: usize = input
        .clone()
        .into_par_iter()
        .map(|x| solve(&x, 24))
        .enumerate()
        .map(|(i, v)| (i + 1) * v)
        .sum();
    println!("{p1}");

    let input_2 = input.iter().take(3).collect::<Vec<_>>();
    let p2: usize = input_2.into_par_iter().map(|x| solve(x, 32)).product();
    println!("{p2}");

    let time = now.elapsed().as_micros();
    eprintln!("Time: {}us", time);
}
